File: app/src/main/java/com/guideapp/ui/views/localdetail/LocalDetailFragment.java
package com.guideapp.ui.views.localdetail;

import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.ShareCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.graphics.Palette;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.BitmapImageViewTarget;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.guideapp.R;
import com.guideapp.utilities.Constants;

public class LocalDetailFragment extends Fragment
        implements LocalDetailContract.View, OnMapReadyCallback, GoogleMap.OnMarkerClickListener {
    private static final String EXTRA_LOCAL_ID = "local_id";

    private double mLatitude;
    private double mLongitude;
    private int mIdImageMarker;

    private CollapsingToolbarLayout mCollapsing;
    private FloatingActionButton mFab;
    private ImageView mImage;
    private LocalDetailContract.Presenter mPresenter;

    private TextView mSubCategoryView;
    private TextView mDescriptionView;
    private TextView mDirectionActionView;
    private TextView mCallActionView;
    private TextView mWebsiteActionView;
    private TextView mAddressView;
    private TextView mPhoneView;

    private SupportMapFragment mSupportMapFragment;
    private FirebaseAnalytics mFirebaseAnalytics;
    private String mTitle;

    public static Fragment newInstance(long id) {
        Fragment fragment = new LocalDetailFragment();
        Bundle bundle = new Bundle();
        bundle.putLong(EXTRA_LOCAL_ID, id);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        mFirebaseAnalytics = FirebaseAnalytics.getInstance(getContext());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_local_detail, container, false);

        setupViews(view);

        return view;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_detail, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        switch (id) {
            case R.id.action_share:
                mPresenter.shareLocal();
                break;

            default:
                return super.onOptionsItemSelected(item);
        }

        return super.onOptionsItemSelected(item);
    }

    private void setupViews(View view) {
        mCollapsing = (CollapsingToolbarLayout) getActivity().findViewById(R.id.collapsingToolbarLayout);
        mImage = (ImageView) getActivity().findViewById(R.id.image);
        mFab = (FloatingActionButton) getActivity().findViewById(R.id.fab);

        mSubCategoryView = (TextView) view.findViewById(R.id.sub_category_text);
        mDescriptionView = (TextView) view.findViewById(R.id.description_text);
        mAddressView = (TextView) view.findViewById(R.id.address_text);
        mPhoneView = (TextView) view.findViewById(R.id.phone_text);

        mDirectionActionView = (TextView) view.findViewById(R.id.direction_action);
        mCallActionView = (TextView) view.findViewById(R.id.call_action);
        mWebsiteActionView = (TextView) view.findViewById(R.id.website_action);

        mFab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mPresenter.saveOrRemoveFavorite();

            }
        });

        mAddressView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mPresenter.loadDirection();

            }
        });

        mDirectionActionView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mPresenter.loadDirection();
            }
        });

        mPhoneView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mPresenter.loadCall();
            }
        });

        mCallActionView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mPresenter.loadCall();
            }
        });

        mWebsiteActionView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mPresenter.loadWebsite();
            }
        });

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        if (mPresenter != null) {
            mPresenter.destroy(getActivity().getSupportLoaderManager());
            mPresenter = null;
        }
    }

    @Override
    public void showFavoriteYes() {
        mFab.setImageDrawable(ContextCompat.getDrawable(getContext(), R.drawable.ic_bookmark_white_24dp));
    }

    @Override
    public void showFavoriteNo() {
        mFab.setImageDrawable(ContextCompat.getDrawable(getContext(), R.drawable.ic_bookmark_border_white_24dp));
    }

    @Override
    public void showSnackbarRemoveFavorite() {
        Snackbar.make(mCollapsing, R.string.title_remove_favorite, Snackbar.LENGTH_SHORT).show();
        sendEventFavorite(Constants.Analytics.REMOVE_FAVORITE);
    }

    @Override
    public void showSnackbarSaveFavorite() {
        Snackbar.make(mCollapsing, R.string.title_save_favorite, Snackbar.LENGTH_SHORT).show();
        sendEventFavorite(Constants.Analytics.SAVE_FAVORITE);
    }

    private void sendEventFavorite(String type) {
        Bundle bundle = new Bundle();

        if (getArguments() != null) {
            long id = getArguments().getLong(EXTRA_LOCAL_ID);
            bundle.putLong(FirebaseAnalytics.Param.ITEM_ID, id);
            bundle.putString(FirebaseAnalytics.Param.ITEM_NAME, mTitle);
            bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, type);
            mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle);
        }
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (getArguments() != null) {
            long id = getArguments().getLong(EXTRA_LOCAL_ID);

            mPresenter = new LocalDetailPresenter(this, id);
            mPresenter.loadLocal(getActivity().getSupportLoaderManager());
        }
    }

    @Override
    public void shareText(String textToShare) {
        String mimeType = "text/plain";
        ShareCompat.IntentBuilder
                .from(this.getActivity())
                .setType(mimeType)
                .setChooserTitle(R.string.app_name)
                .setText(textToShare)
                .startChooser();
    }

    @Override
    public void showTitle(String description) {
        mTitle = description;
        mCollapsing.setTitle(description);
        mCollapsing.setExpandedTitleColor(ContextCompat.getColor(getContext(), R.color.white));
    }

    @Override
    public void showImage(String imagePath) {
        Glide.with(this)
                .load(imagePath)
                .asBitmap()
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(new BitmapImageViewTarget(mImage) {
                    @Override
                    public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {
                        super.onResourceReady(resource, glideAnimation);
                        mImage.setBackgroundResource(R.drawable.action_background_bottom);
                        Palette.from(resource).generate(new Palette.PaletteAsyncListener() {
                            @Override
                            public void onGenerated(Palette palette) {
                                applyPalette(palette);
                            }
                        });
                    }
                });
    }

    @Override
    public void showCategory(String text) {
        mSubCategoryView.setText(text);
    }

    @Override
    public void showWebSiteAction() {
        mWebsiteActionView.setVisibility(View.VISIBLE);
    }

    @Override
    public void showDirectionAction() {
        mDirectionActionView.setVisibility(View.VISIBLE);
    }

    @Override
    public void showCallAction() {
        mCallActionView.setVisibility(View.VISIBLE);
    }

    @Override
    public void showCall(String phone) {
        mPhoneView.setVisibility(View.VISIBLE);
        mPhoneView.setText(phone);
    }

    @Override
    public void showDetail(String description) {
        mDescriptionView.setVisibility(View.VISIBLE);
        mDescriptionView.setText(description);
    }

    @Override
    public void showAddress(String address) {
        mAddressView.setVisibility(View.VISIBLE);
        mAddressView.setText(address);
    }

    @Override
    public void showMap(double latitude, double longitude, int idImageMarker) {
        mLatitude = latitude;
        mLongitude = longitude;
        mIdImageMarker = idImageMarker;
        setUpMapIfNeeded();

        getActivity().supportStartPostponedEnterTransition();
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        LatLng center = new LatLng(mLatitude, mLongitude);
        googleMap.moveCamera(CameraUpdateFactory.newLatLng(center));

        MarkerOptions markerOptions = new MarkerOptions()
                .position(center)
                .icon(BitmapDescriptorFactory.fromResource(mIdImageMarker));

        googleMap.clear();
        googleMap.addMarker(markerOptions);
        googleMap.getUiSettings().setMapToolbarEnabled(false);

        googleMap.setOnMarkerClickListener(this);
        googleMap.setOnMapClickListener(new GoogleMap.OnMapClickListener() {
            @Override
            public void onMapClick(LatLng latLng) {
                mPresenter.loadDirection();

            }
        });
    }

    private void setUpMapIfNeeded() {
        if (mSupportMapFragment == null) {
            FragmentManager fm = getChildFragmentManager();
            mSupportMapFragment = ((SupportMapFragment) fm.findFragmentById(R.id.map));
            mSupportMapFragment.getMapAsync(this);
        }
    }

    private void applyPalette(Palette palette) {
        int primary = ContextCompat.getColor(getContext(), R.color.blue_grey_500);
        mCollapsing.setContentScrimColor(palette.getVibrantColor(primary));
        mCollapsing.setStatusBarScrimColor(palette.getVibrantColor(primary));
    }

    @Override
    public void dialPhoneNumber(String number) {
        Intent intent = new Intent(Intent.ACTION_DIAL);
        intent.setData(Uri.parse("tel:" + number));
        if (intent.resolveActivity(getActivity().getPackageManager()) != null) {
            startActivity(intent);
        }
    }

    @Override
    public void openPage(String url) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(url));
        if (intent.resolveActivity(getActivity().getPackageManager()) != null) {
            startActivity(intent);
        }
    }

    @Override
    public void openDirection(String description, String latLng) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse("geo:" + latLng + "?q=" + latLng + "(" + description + ")"));
        if (intent.resolveActivity(getActivity().getPackageManager()) != null) {
            startActivity(intent);
        }
    }

    @Override
    public boolean onMarkerClick(Marker marker) {
        mPresenter.loadDirection();
        return false;
    }
}

File: app/src/main/java/com/guideapp/ui/views/localdetail/LocalDetailFragment.kt
package com.guideapp.ui.views.localdetail

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.CollapsingToolbarLayout
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v4.app.ShareCompat
import android.support.v4.content.ContextCompat
import android.support.v7.graphics.Palette
import android.view.*
import android.widget.ImageView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.animation.GlideAnimation
import com.bumptech.glide.request.target.BitmapImageViewTarget
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import com.google.firebase.analytics.FirebaseAnalytics
import com.guideapp.R
import com.guideapp.utilities.Constants
import kotlinx.android.synthetic.main.fragment_local_detail.*

class LocalDetailFragment : Fragment(), LocalDetailContract.View, OnMapReadyCallback, GoogleMap.OnMarkerClickListener {
    private var mLatitude: Double = 0.toDouble()
    private var mLongitude: Double = 0.toDouble()
    private var mIdImageMarker: Int = 0

    private var mCollapsing: CollapsingToolbarLayout? = null
    private var mFab: FloatingActionButton? = null
    private var mImage: ImageView? = null
    private var mPresenter: LocalDetailContract.Presenter? = null

    private var mSupportMapFragment: SupportMapFragment? = null
    private var mFirebaseAnalytics: FirebaseAnalytics? = null
    private var mTitle: String? = null

    @SuppressLint("MissingPermission")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        mFirebaseAnalytics = FirebaseAnalytics.getInstance(context)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater!!.inflate(R.layout.fragment_local_detail, container, false)
        setupViewActivity()
        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setupViews()
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater?.inflate(R.menu.menu_detail, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        val id = item?.itemId
        when (id) {
            R.id.action_share -> mPresenter?.shareLocal()
            else -> return super.onOptionsItemSelected(item)
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setupViewActivity() {
        mCollapsing = activity.findViewById(R.id.collapsingToolbarLayout) as CollapsingToolbarLayout
        mImage = activity.findViewById(R.id.image) as ImageView
        mFab = activity.findViewById(R.id.fab) as FloatingActionButton
        mFab?.setOnClickListener { mPresenter?.saveOrRemoveFavorite() }
    }

    private fun setupViews() {
        address_text?.setOnClickListener { mPresenter?.loadDirection() }
        phone_text?.setOnClickListener { mPresenter?.loadCall() }
        direction_action?.setOnClickListener { mPresenter?.loadDirection() }
        call_action?.setOnClickListener { mPresenter?.loadCall() }
        website_action?.setOnClickListener { mPresenter?.loadWebsite() }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mPresenter?.destroy(activity.supportLoaderManager)
        mPresenter = null
    }

    override fun showFavoriteYes() {
        mFab?.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_bookmark_white_24dp))
    }

    override fun showFavoriteNo() {
        mFab?.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_bookmark_border_white_24dp))
    }

    override fun showSnackbarRemoveFavorite() {
        mCollapsing?.let { collapsing ->
            Snackbar.make(collapsing, R.string.title_remove_favorite, Snackbar.LENGTH_SHORT).show()
        }
        sendEventFavorite(Constants.Analytics.REMOVE_FAVORITE)
    }

    override fun showSnackbarSaveFavorite() {
        mCollapsing?.let { collapsing ->
            Snackbar.make(collapsing, R.string.title_save_favorite, Snackbar.LENGTH_SHORT).show()
        }
        sendEventFavorite(Constants.Analytics.SAVE_FAVORITE)
    }

    private fun sendEventFavorite(type: String) {
        val bundle = Bundle()
        if (arguments != null) {
            val id = arguments.getLong(EXTRA_LOCAL_ID)
            bundle.putLong(FirebaseAnalytics.Param.ITEM_ID, id)
            bundle.putString(FirebaseAnalytics.Param.ITEM_NAME, mTitle)
            bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, type)
            mFirebaseAnalytics?.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle)
        }
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        arguments?.let { argument ->
            val id = argument.getLong(EXTRA_LOCAL_ID)
            mPresenter = LocalDetailPresenter(this, id)
            mPresenter?.loadLocal(activity.supportLoaderManager)
        }
    }

    override fun shareText(textToShare: String) {
        val mimeType = "text/plain"
        ShareCompat.IntentBuilder
                .from(this.activity)
                .setType(mimeType)
                .setChooserTitle(R.string.app_name)
                .setText(textToShare)
                .startChooser()
    }

    override fun showTitle(description: String) {
        mTitle = description
        mCollapsing?.title = description
        mCollapsing?.setExpandedTitleColor(ContextCompat.getColor(context, R.color.white))
    }

    override fun showImage(imagePath: String) {
        Glide.with(this)
                .load(imagePath)
                .asBitmap()
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(object : BitmapImageViewTarget(mImage) {
                    override fun onResourceReady(resource: Bitmap, glideAnimation: GlideAnimation<in Bitmap>?) {
                        super.onResourceReady(resource, glideAnimation)
                        mImage?.setBackgroundResource(R.drawable.action_background_bottom)
                        Palette.from(resource).generate { palette -> applyPalette(palette) }
                    }
                })
    }

    override fun showCategory(text: String) {
        sub_category_text?.text = text
    }

    override fun showWebSiteAction() {
        website_action?.visibility = View.VISIBLE
    }

    override fun showDirectionAction() {
        direction_action?.visibility = View.VISIBLE
    }

    override fun showCallAction() {
        call_action?.visibility = View.VISIBLE
    }

    override fun showCall(phone: String) {
        phone_text?.visibility = View.VISIBLE
        phone_text?.text = phone
    }

    override fun showDetail(description: String) {
        description_text?.visibility = View.VISIBLE
        description_text?.text = description
    }

    override fun showAddress(address: String) {
        address_text?.visibility = View.VISIBLE
        address_text?.text = address
    }

    override fun showMap(latitude: Double, longitude: Double, idImageMarker: Int) {
        mLatitude = latitude
        mLongitude = longitude
        mIdImageMarker = idImageMarker
        setUpMapIfNeeded()
        activity.supportStartPostponedEnterTransition()
    }

    override fun onMapReady(googleMap: GoogleMap) {
        val center = LatLng(mLatitude, mLongitude)
        googleMap.moveCamera(CameraUpdateFactory.newLatLng(center))

        val markerOptions = MarkerOptions()
                .position(center)
                .icon(BitmapDescriptorFactory.fromResource(mIdImageMarker))

        googleMap.clear()
        googleMap.addMarker(markerOptions)
        googleMap.uiSettings.isMapToolbarEnabled = false

        googleMap.setOnMarkerClickListener(this)
        googleMap.setOnMapClickListener { mPresenter?.loadDirection() }
    }

    private fun setUpMapIfNeeded() {
        if (mSupportMapFragment == null) {
            val fm = childFragmentManager
            mSupportMapFragment = fm.findFragmentById(R.id.map) as SupportMapFragment
            mSupportMapFragment?.getMapAsync(this)
        }
    }

    private fun applyPalette(palette: Palette) {
        val primary = ContextCompat.getColor(context, R.color.blue_grey_500)
        mCollapsing?.setContentScrimColor(palette.getVibrantColor(primary))
        mCollapsing?.setStatusBarScrimColor(palette.getVibrantColor(primary))
    }

    override fun dialPhoneNumber(number: String) {
        val intent = Intent(Intent.ACTION_DIAL)
        intent.data = Uri.parse("tel:" + number)
        if (intent.resolveActivity(activity.packageManager) != null) {
            startActivity(intent)
        }
    }

    override fun openPage(url: String) {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.data = Uri.parse(url)
        if (intent.resolveActivity(activity.packageManager) != null) {
            startActivity(intent)
        }
    }

    override fun openDirection(description: String, latLng: String) {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.data = Uri.parse("geo:$latLng?q=$latLng($description)")
        if (intent.resolveActivity(activity.packageManager) != null) {
            startActivity(intent)
        }
    }

    override fun onMarkerClick(marker: Marker): Boolean {
        mPresenter?.loadDirection()
        return false
    }

    companion object {
        private val EXTRA_LOCAL_ID = "local_id"
        fun newInstance(id: Long): Fragment {
            val fragment = LocalDetailFragment()
            val bundle = Bundle()
            bundle.putLong(EXTRA_LOCAL_ID, id)
            fragment.arguments = bundle
            return fragment
        }
    }
}
